File: ios/Halogen/Features/Player/PlayerSheet.swift
import SwiftUI

/// The full player: "Up next" preview up top, artwork floating in the middle,
/// then title, scrubber (with chapter ticks), transport with previous/next
/// episode, chapters menu, rate, sleep timer and auto-advance quick toggle
/// anchored to the bottom — the web's NowPlayingScreen surface.
struct PlayerSheet: View {
    @Bindable var player: PlayerModel
    let core: HalogenCore

    @State private var scrubbing = false
    @State private var scrubValue: Double = 0

    var body: some View {
        VStack(spacing: 24) {
            Capsule()
                .fill(.tertiary)
                .frame(width: 36, height: 5)
                .padding(.top, 10)

            if let episode = player.current {
                if let next = player.upNext {
                    upNextRow(next)
                }

                Spacer(minLength: 0)

                Artwork(url: core.episodeArtURL(episode, small: false), size: 260)

                Spacer(minLength: 0)

                VStack(spacing: 4) {
                    Text(episode.title)
                        .font(.headline)
                        .multilineTextAlignment(.center)
                        .lineLimit(2)
                        .accessibilityIdentifier("player-title")
                    if let podcast = episode.podcast?.title {
                        Text(podcast).font(.subheadline).foregroundStyle(.secondary)
                    }
                    if let failure = player.failureMessage {
                        // Tappable — retry routes through the full source
                        // rebuild (toggle() handles the failed-item case).
                        Button {
                            player.toggle()
                        } label: {
                            Label(failure, systemImage: "arrow.clockwise.circle.fill")
                                .font(.caption)
                                .foregroundStyle(.orange)
                                .multilineTextAlignment(.center)
                        }
                        .buttonStyle(.plain)
                        .accessibilityIdentifier("player-retry")
                    } else if player.preparing {
                        HStack(spacing: 6) {
                            ProgressView().controlSize(.small)
                            Text(player.preparingLabel)
                        }
                        .font(.caption)
                        .foregroundStyle(.secondary)
                    } else if player.streaming {
                        Label("Streaming", systemImage: "dot.radiowaves.left.and.right")
                            .font(.caption)
                            .foregroundStyle(.secondary)
                    }
                }
                .padding(.horizontal, 24)

                scrubber

                transportRow

                controlsRow
            } else {
                Spacer()
            }
        }
        .padding(.bottom, 12)
        .presentationDetents([.large])
    }

    /// "Up next" preview — the continuation neighbor (queue, or the playlist
    /// played from), the web's top-right UpNext surface. Tap skips to it.
    private func upNextRow(_ next: EpisodeData) -> some View {
        Button {
            player.playNextEpisode()
        } label: {
            HStack(spacing: 10) {
                Artwork(url: core.episodeArtURL(next), size: 36)
                VStack(alignment: .leading, spacing: 1) {
                    Text("Up next")
                        .font(.caption2.weight(.semibold))
                        .textCase(.uppercase)
                        .foregroundStyle(.secondary)
                    Text(next.title)
                        .font(.footnote.weight(.medium))
                        .lineLimit(1)
                    if let podcast = next.podcast?.title {
                        Text(podcast).font(.caption2).foregroundStyle(.secondary)
                            .lineLimit(1)
                    }
                }
                Spacer()
                Image(systemName: "forward.end.fill")
                    .font(.caption)
                    .foregroundStyle(.secondary)
            }
            .padding(10)
            .background(RoundedRectangle(cornerRadius: 12).fill(Color(.secondarySystemFill)))
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .accessibilityIdentifier("player-up-next")
        .padding(.horizontal, 24)
    }

    private var transportRow: some View {
        HStack(spacing: 32) {
            Button {
                player.playPreviousEpisode()
            } label: {
                Image(systemName: "backward.end.fill").font(.title2)
            }
            .disabled(!player.hasPrevious)
            Button {
                player.skip(-player.skipBackSecs)
            } label: {
                Image(systemName: "gobackward").font(.title)
            }
            Button {
                player.toggle()
            } label: {
                if player.buffering {
                    // Rebuffering after a dry buffer — a play/pause
                    // glyph over silence misreads as "playing fine".
                    ProgressView()
                        .controlSize(.large)
                        .frame(width: 64, height: 64)
                } else {
                    Image(
                        systemName: player.isPlaying
                            ? "pause.circle.fill" : "play.circle.fill"
                    )
                    .font(.system(size: 64))
                }
            }
            Button {
                player.skip(player.skipForwardSecs)
            } label: {
                Image(systemName: "goforward").font(.title)
            }
            Button {
                player.playNextEpisode()
            } label: {
                Image(systemName: "forward.end.fill").font(.title2)
            }
            .accessibilityIdentifier("player-next")
            .disabled(!player.hasNext)
        }
        .buttonStyle(.plain)
    }

    private var controlsRow: some View {
        HStack(spacing: 16) {
            if !player.chapters.isEmpty {
                chaptersMenu
            }

            Menu {
                ForEach(ClientPrefs.playbackRates, id: \.self) { rate in
                    Button(String(format: "%g×", rate)) {
                        player.setRate(rate)
                    }
                }
            } label: {
                Text(String(format: "%g×", player.rate))
                    .font(.callout.weight(.semibold))
                    .padding(.horizontal, 14)
                    .padding(.vertical, 6)
                    .background(Capsule().fill(Color(.secondarySystemFill)))
            }

            // Auto-advance quick toggle — same pref as Settings
            // (web: the now-playing "A" toggle persists immediately).
            Button {
                core.models?.prefs.update { $0.autoAdvance.toggle() }
            } label: {
                Image(
                    systemName: autoAdvance
                        ? "a.circle.fill" : "a.circle"
                )
                .font(.title3)
                .foregroundStyle(autoAdvance ? Color.accentColor : .secondary)
                .padding(.vertical, 6)
            }

            Menu {
                Button("Off") { player.setSleepTimer(minutes: nil) }
                ForEach(sleepMenuMinutes, id: \.self) { m in
                    Button("\(m) min") { player.setSleepTimer(minutes: m) }
                }
                Button("End of episode") { player.setSleepTimer(minutes: 0) }
            } label: {
                Label(sleepLabel, systemImage: "moon.zzz")
                    .font(.callout.weight(.semibold))
                    .padding(.horizontal, 14)
                    .padding(.vertical, 6)
                    .background(Capsule().fill(Color(.secondarySystemFill)))
            }
        }
    }

    private var autoAdvance: Bool {
        core.models?.prefs.prefs.autoAdvance ?? true
    }

    /// The quick sleep durations, always including the configured default
    /// (Settings → Playback) so it's armable from here.
    private var sleepMenuMinutes: [Int] {
        let preset = core.models?.prefs.prefs.defaultSleepMinutes ?? 30
        return Array(Set([5, 15, 30, 60, preset])).sorted()
    }

    /// Chapter picker — selecting a marker seeks to it (web: the drop-up
    /// chapters menu). The active marker is checked.
    private var chaptersMenu: some View {
        Menu {
            ForEach(Array(player.chapters.enumerated()), id: \.element.id) { index, chapter in
                Button {
                    player.seek(to: Double(chapter.starts_at_secs))
                } label: {
                    if player.activeChapterIndex == index {
                        Label(
                            "\(Self.time(Double(chapter.starts_at_secs)))  \(chapter.title)",
                            systemImage: "checkmark")
                    } else {
                        Text("\(Self.time(Double(chapter.starts_at_secs)))  \(chapter.title)")
                    }
                }
            }
        } label: {
            Image(systemName: "list.bullet")
                .font(.title3)
                .padding(.vertical, 6)
        }
    }

    private var scrubber: some View {
        VStack(spacing: 4) {
            // Current chapter — updates as playback crosses each marker
            // (web: the title above the seek bar).
            if let idx = player.activeChapterIndex {
                Text("\(idx + 1). \(player.chapters[idx].title)")
                    .font(.caption)
                    .foregroundStyle(.secondary)
                    .lineLimit(1)
                    .frame(maxWidth: .infinity, alignment: .leading)
            }
            Slider(
                value: Binding(
                    get: { scrubbing ? scrubValue : player.position },
                    set: { scrubValue = $0 }
                ),
                in: 0...max(player.duration, 1),
                onEditingChanged: { editing in
                    if editing {
                        scrubValue = player.position
                        scrubbing = true
                    } else {
                        scrubbing = false
                        player.seek(to: scrubValue)
                    }
                }
            )
            .overlay {
                // Chapter tick marks — decorative (never intercept touches),
                // the web's absolute-positioned lines over the range input.
                if player.duration > 0, player.chapters.count > 1 {
                    GeometryReader { geo in
                        ForEach(player.chapters, id: \.id) { chapter in
                            let frac = Double(chapter.starts_at_secs) / player.duration
                            if frac > 0, frac < 1 {
                                Rectangle()
                                    .fill(.secondary.opacity(0.6))
                                    .frame(width: 1, height: 8)
                                    .position(
                                        x: geo.size.width * frac,
                                        y: geo.size.height / 2)
                            }
                        }
                    }
                    .allowsHitTesting(false)
                }
            }
            HStack {
                Text(Self.time(scrubbing ? scrubValue : player.position))
                Spacer()
                Text("-" + Self.time(max(0, player.duration - (scrubbing ? scrubValue : player.position))))
            }
            .font(.caption.monospacedDigit())
            .foregroundStyle(.secondary)
        }
        .padding(.horizontal, 24)
    }

    private var sleepLabel: String {
        if player.sleepAtEpisodeEnd { return "Sleep: end" }
        if let mins = player.sleepRemainingMinutes {
            return "Sleep: \(mins)m"
        }
        return "Sleep"
    }

    private static func time(_ seconds: Double) -> String {
        let s = Int(seconds)
        return s >= 3600
            ? String(format: "%d:%02d:%02d", s / 3600, (s % 3600) / 60, s % 60)
            : String(format: "%d:%02d", s / 60, s % 60)
    }
}

File: android/app/src/main/kotlin/org/fgsec/halogen/features/player/PlayerSheet.kt
@file:OptIn(ExperimentalMaterial3Api::class)

package org.fgsec.halogen.features.player

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Slider
import androidx.compose.material3.SliderDefaults
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.drawWithContent
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import kotlin.math.max
import org.fgsec.halogen.components.Artwork
import org.fgsec.halogen.components.halogenIcon
import org.fgsec.halogen.core.ClientPrefs
import org.fgsec.halogen.core.HalogenCore
import org.fgsec.halogen.wire.EpisodeData

/// The full player: up-next preview, floating artwork, title, scrubber with chapter
/// ticks, transport, then chapters/rate/sleep/auto-advance anchored at the bottom —
/// the web's NowPlayingScreen surface. Mirrors ios Features/Player/PlayerSheet.swift.
@Composable
fun PlayerSheet(player: PlayerModel, core: HalogenCore, onDismiss: () -> Unit) {
    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true),
    ) {
        val episode = player.current
        if (episode == null) {
            Spacer(Modifier.height(200.dp))
            return@ModalBottomSheet
        }
        // Everything below the art is fixed-height; the art takes what is
        // left. A fixed 260dp pushed the transport off the bottom at larger
        // UI sizes, and the sheet does not scroll by itself.
        BoxWithConstraints(Modifier.fillMaxSize()) {
            val art = (maxHeight * 0.30f).coerceIn(112.dp, 260.dp)
                .coerceAtMost(maxWidth - 96.dp)
            Column(
                Modifier
                    .fillMaxSize()
                    .verticalScroll(rememberScrollState())
                    .navigationBarsPadding()
                    .padding(bottom = 12.dp),
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.spacedBy(12.dp),
            ) {
                player.upNext?.let { next -> UpNextRow(next, core, player) }

                Artwork(url = core.episodeArtUrl(episode, small = false), size = art)

                TitleCluster(player, episode)

                Scrubber(player)

                TransportRow(player)

                ControlsRow(player, core)
            }
        }
    }
}

/// "Up next" preview — the continuation neighbor (queue, or the playlist
/// played from), the web's top-right UpNext surface. Tap skips to it.
@Composable
private fun UpNextRow(next: EpisodeData, core: HalogenCore, player: PlayerModel) {
    Surface(
        shape = RoundedCornerShape(12.dp),
        color = MaterialTheme.colorScheme.surfaceContainerHigh,
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = 24.dp)
            .testTag("player-up-next"),
        onClick = { player.playNextEpisode() },
    ) {
        Row(
            Modifier.padding(10.dp),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(10.dp),
        ) {
            Artwork(url = core.episodeArtUrl(next), size = 36.dp)
            Column(Modifier.weight(1f)) {
                Text(
                    "UP NEXT",
                    style = MaterialTheme.typography.labelSmall,
                    fontWeight = FontWeight.SemiBold,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
                Text(
                    next.title,
                    style = MaterialTheme.typography.bodySmall,
                    fontWeight = FontWeight.Medium,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                )
                next.podcast?.title?.let { podcast ->
                    Text(
                        podcast,
                        style = MaterialTheme.typography.labelSmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis,
                    )
                }
            }
            Icon(
                halogenIcon("forward.end.fill"),
                contentDescription = null,
                tint = MaterialTheme.colorScheme.onSurfaceVariant,
                modifier = Modifier.size(16.dp),
            )
        }
    }
}

@Composable
private fun TitleCluster(player: PlayerModel, episode: EpisodeData) {
    Column(
        Modifier.padding(horizontal = 24.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(4.dp),
    ) {
        Text(
            episode.title,
            style = MaterialTheme.typography.titleMedium,
            textAlign = TextAlign.Center,
            maxLines = 2,
            overflow = TextOverflow.Ellipsis,
            modifier = Modifier.testTag("player-title"),
        )
        episode.podcast?.title?.let { podcast ->
            Text(
                podcast,
                style = MaterialTheme.typography.labelMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }
        val failure = player.failureMessage
        when {
            failure != null ->
                // Tappable — retry routes through the full source rebuild
                // (toggle() handles the failed-item case).
                Row(
                    Modifier
                        .clickable { player.toggle() }
                        .testTag("player-retry"),
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(6.dp),
                ) {
                    Icon(
                        halogenIcon("arrow.clockwise.circle.fill"),
                        contentDescription = null,
                        tint = MaterialTheme.colorScheme.error,
                        modifier = Modifier.size(16.dp),
                    )
                    Text(
                        failure,
                        style = MaterialTheme.typography.labelSmall,
                        color = MaterialTheme.colorScheme.error,
                        textAlign = TextAlign.Center,
                    )
                }
            player.preparing ->
                Row(
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(6.dp),
                ) {
                    CircularProgressIndicator(Modifier.size(14.dp), strokeWidth = 2.dp)
                    Text(
                        player.preparingLabel,
                        style = MaterialTheme.typography.labelSmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                    )
                }
            player.streaming ->
                Row(
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(4.dp),
                ) {
                    Icon(
                        halogenIcon("dot.radiowaves.left.and.right"),
                        contentDescription = null,
                        tint = MaterialTheme.colorScheme.onSurfaceVariant,
                        modifier = Modifier.size(14.dp),
                    )
                    Text(
                        "Streaming",
                        style = MaterialTheme.typography.labelSmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                    )
                }
        }
    }
}

@Composable
private fun Scrubber(player: PlayerModel) {
    var scrubbing by remember { mutableStateOf(false) }
    var scrubValue by remember { mutableFloatStateOf(0f) }
    val shown = if (scrubbing) scrubValue.toDouble() else player.position
    val tickColor = MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = 0.6f)

    Column(
        Modifier
            .fillMaxWidth()
            .padding(horizontal = 24.dp),
        verticalArrangement = Arrangement.spacedBy(4.dp),
    ) {
        // Current chapter — updates as playback crosses each marker
        // (web: the title above the seek bar).
        player.activeChapterIndex?.let { idx ->
            Text(
                "${idx + 1}. ${player.chapters[idx].title}",
                style = MaterialTheme.typography.labelSmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
                modifier = Modifier.fillMaxWidth(),
            )
        }
        val duration = max(player.duration, 1.0)
        val chapters = player.chapters
        Slider(
            value = shown.toFloat().coerceIn(0f, duration.toFloat()),
            onValueChange = {
                scrubbing = true
                scrubValue = it
            },
            onValueChangeFinished = {
                scrubbing = false
                player.seek(scrubValue.toDouble())
            },
            valueRange = 0f..duration.toFloat(),
            modifier = Modifier
                .fillMaxWidth()
                // Chapter tick marks — decorative, the web's absolute-
                // positioned lines over the range input.
                .drawWithContent {
                    drawContent()
                    if (player.duration > 0 && chapters.size > 1) {
                        for (chapter in chapters) {
                            val frac = chapter.starts_at_secs / player.duration
                            if (frac > 0 && frac < 1) {
                                val x = size.width * frac.toFloat()
                                drawLine(
                                    color = tickColor,
                                    start = Offset(x, size.height / 2 - 4.dp.toPx()),
                                    end = Offset(x, size.height / 2 + 4.dp.toPx()),
                                    strokeWidth = 1.dp.toPx(),
                                )
                            }
                        }
                    }
                },
            track = { state ->
                SliderDefaults.Track(sliderState = state, drawStopIndicator = null)
            },
        )
        Row(Modifier.fillMaxWidth()) {
            Text(
                formatTime(shown),
                style = MaterialTheme.typography.labelSmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
            Spacer(Modifier.weight(1f))
            Text(
                "-" + formatTime(max(0.0, player.duration - shown)),
                style = MaterialTheme.typography.labelSmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }
    }
}

@Composable
private fun TransportRow(player: PlayerModel) {
    Row(
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(16.dp),
    ) {
        IconButton(onClick = { player.playPreviousEpisode() }, enabled = player.hasPrevious) {
            Icon(
                halogenIcon("backward.end.fill"),
                contentDescription = "Previous episode",
                modifier = Modifier.size(28.dp),
            )
        }
        IconButton(onClick = { player.skip(-player.skipBackSecs) }) {
            Icon(
                halogenIcon("gobackward"),
                contentDescription = "Skip back",
                modifier = Modifier.size(32.dp),
            )
        }
        // Filled circle, as on iOS — a bare glyph read as one more icon in
        // the row rather than the primary control.
        Surface(
            shape = CircleShape,
            color = MaterialTheme.colorScheme.onSurface,
            onClick = { player.toggle() },
            modifier = Modifier.size(64.dp),
        ) {
            Box(contentAlignment = Alignment.Center) {
                if (player.buffering) {
                    CircularProgressIndicator(
                        Modifier.size(32.dp),
                        color = MaterialTheme.colorScheme.surface,
                    )
                } else {
                    Icon(
                        halogenIcon(if (player.isPlaying) "pause.fill" else "play.fill"),
                        contentDescription = if (player.isPlaying) "Pause" else "Play",
                        tint = MaterialTheme.colorScheme.surface,
                        modifier = Modifier.size(34.dp),
                    )
                }
            }
        }
        IconButton(onClick = { player.skip(player.skipForwardSecs) }) {
            Icon(
                halogenIcon("goforward"),
                contentDescription = "Skip forward",
                modifier = Modifier.size(32.dp),
            )
        }
        IconButton(
            onClick = { player.playNextEpisode() },
            enabled = player.hasNext,
            modifier = Modifier.testTag("player-next"),
        ) {
            Icon(
                halogenIcon("forward.end.fill"),
                contentDescription = "Next episode",
                modifier = Modifier.size(28.dp),
            )
        }
    }
}

@Composable
private fun ControlsRow(player: PlayerModel, core: HalogenCore) {
    Row(
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(16.dp),
    ) {
        if (player.chapters.isNotEmpty()) {
            ChaptersMenu(player)
        }

        RateMenu(player)

        // Auto-advance quick toggle — same pref as Settings (web: the
        // now-playing "A" toggle persists immediately).
        val autoAdvance = core.models?.prefs?.prefs?.autoAdvance ?: true
        IconButton(
            onClick = { core.models?.prefs?.update { it.copy(autoAdvance = !it.autoAdvance) } }
        ) {
            Box(
                Modifier
                    .size(24.dp)
                    .then(
                        if (autoAdvance)
                            Modifier.background(MaterialTheme.colorScheme.primary, CircleShape)
                        else
                            Modifier.border(
                                1.5.dp, MaterialTheme.colorScheme.onSurfaceVariant, CircleShape
                            )
                    ),
                contentAlignment = Alignment.Center,
            ) {
                Text(
                    "A",
                    style = MaterialTheme.typography.labelSmall,
                    fontWeight = FontWeight.Bold,
                    color =
                        if (autoAdvance) MaterialTheme.colorScheme.onPrimary
                        else MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }
        }

        SleepMenu(player, core)
    }
}

/// Chapter picker — selecting a marker seeks to it (web: the drop-up
/// chapters menu). The active marker is checked.
@Composable
private fun ChaptersMenu(player: PlayerModel) {
    var expanded by remember { mutableStateOf(false) }
    Box {
        IconButton(onClick = { expanded = true }) {
            Icon(halogenIcon("list.bullet"), contentDescription = "Chapters")
        }
        DropdownMenu(expanded = expanded, onDismissRequest = { expanded = false }) {
            player.chapters.forEachIndexed { index, chapter ->
                DropdownMenuItem(
                    text = {
                        Text("${formatTime(chapter.starts_at_secs.toDouble())}  ${chapter.title}")
                    },
                    leadingIcon =
                        if (player.activeChapterIndex == index) {
                            {
                                Icon(halogenIcon("checkmark"), contentDescription = null)
                            }
                        } else null,
                    onClick = {
                        expanded = false
                        player.seek(chapter.starts_at_secs.toDouble())
                    },
                )
            }
        }
    }
}

@Composable
private fun RateMenu(player: PlayerModel) {
    var expanded by remember { mutableStateOf(false) }
    Box {
        Chip(label = rateLabel(player.rate), onClick = { expanded = true })
        DropdownMenu(expanded = expanded, onDismissRequest = { expanded = false }) {
            for (rate in ClientPrefs.playbackRates) {
                DropdownMenuItem(
                    text = { Text(rateLabel(rate)) },
                    onClick = {
                        expanded = false
                        player.applyRate(rate)
                    },
                )
            }
        }
    }
}

@Composable
private fun SleepMenu(player: PlayerModel, core: HalogenCore) {
    var expanded by remember { mutableStateOf(false) }
    // The quick sleep durations, always including the configured default
    // (Settings → Playback) so it's armable from here.
    val preset = core.models?.prefs?.prefs?.defaultSleepMinutes ?: 30
    val minutes = (listOf(5, 15, 30, 60) + preset).distinct().sorted()
    Box {
        Chip(label = sleepLabel(player), icon = "moon.zzz", onClick = { expanded = true })
        DropdownMenu(expanded = expanded, onDismissRequest = { expanded = false }) {
            DropdownMenuItem(
                text = { Text("Off") },
                onClick = {
                    expanded = false
                    player.setSleepTimer(null)
                },
            )
            for (m in minutes) {
                DropdownMenuItem(
                    text = { Text("$m min") },
                    onClick = {
                        expanded = false
                        player.setSleepTimer(m)
                    },
                )
            }
            DropdownMenuItem(
                text = { Text("End of episode") },
                onClick = {
                    expanded = false
                    player.setSleepTimer(0)
                },
            )
        }
    }
}

@Composable
private fun Chip(label: String, icon: String? = null, onClick: () -> Unit) {
    Surface(
        shape = CircleShape,
        color = MaterialTheme.colorScheme.surfaceContainerHigh,
        onClick = onClick,
    ) {
        Row(
            Modifier.padding(horizontal = 14.dp, vertical = 6.dp),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(6.dp),
        ) {
            if (icon != null) {
                Icon(
                    halogenIcon(icon),
                    contentDescription = null,
                    modifier = Modifier.size(16.dp),
                )
            }
            Text(
                label,
                style = MaterialTheme.typography.labelMedium,
                fontWeight = FontWeight.SemiBold,
            )
        }
    }
}

private fun sleepLabel(player: PlayerModel): String {
    if (player.sleepAtEpisodeEnd) return "Sleep: end"
    player.sleepRemainingMinutes?.let { return "Sleep: ${it}m" }
    return "Sleep"
}

/// Swift's %g rate formatting: "1×", "1.25×" — no trailing zeros.
private fun rateLabel(rate: Float): String {
    val text =
        if (rate == rate.toInt().toFloat()) rate.toInt().toString()
        else rate.toString().trimEnd('0').trimEnd('.')
    return "$text×"
}

private fun formatTime(seconds: Double): String {
    val s = seconds.toInt().coerceAtLeast(0)
    return if (s >= 3600) "%d:%02d:%02d".format(s / 3600, (s % 3600) / 60, s % 60)
    else "%d:%02d".format(s / 60, s % 60)
}
